ba/org.osate.ba/src/org/osate/ba/texteditor/XtextAadlBaHighlighter.java

```diff
@@ -58,4 +60,24 @@ public void addToHighlighting(int annexOffset, int relativeOffset, int length, S
               id));
   }
   
+  private XtextAadlBaHighlighter() {}
+
+  private static Map<AnnexSubclause, XtextAadlBaHighlighter> _highlighterPerAnnex = 
+      new HashMap<AnnexSubclause, XtextAadlBaHighlighter>();
+  
+  public static XtextAadlBaHighlighter getHighlighter(AnnexSubclause as)
+  {
+    if(_highlighterPerAnnex.get(as)==null)
+    {
+      XtextAadlBaHighlighter ht = new XtextAadlBaHighlighter();
+      _highlighterPerAnnex.put(as, ht);
```

Comment:
Why are highlighters stored in a has map? Each run of the parser creates new AnnexSubclause objects, so new highlighters will be created for each parser run.

Reply:
The reason for using a map is the following: text to highlight in the BA subclause is identified during parsing (when the parser is executed from AnnexParserAgent class) but syntax highlighting is processed latter on (when highlightAnnexSubclause is called in class AnnexHighlighterProxy). So we need to save the locations to highlight during parsing and retreive them during highlighting.

The only shared (and stable) information between these methods invocation is the AnnexSubclause (or AnnexLibrary) object. This is why we store the locations to highlight in a map where the AnnexSubclause object is the key. This is also the reason why we store the current AnnexSubclause object in a global variable before starting the parsing. In AnnexParserAgent:

`if(defaultAnnexSection instanceof AnnexSubclause)
					AnnexUtil.setCurrentAnnexSubclause((AnnexSubclause) defaultAnnexSection);
`

I agree it is not very clean but given the methods provided by OSATE for syntax hihglighting of ANTLR based annexes I did not find a better solution so far.